File: 105.从前序和中序遍历序列构造二叉树/buildTree.go
package problem0105

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 中序inorder: 左中右 (9)(3)(15 20 7), 3的索引值作为pivot
// 前序preorder: 中左右 (3)(9)(20 15 7) 第一个元素一定是新建二叉树的根节点
// 关键特点:
// 对于inorder[:pivot], 其值分布在preorder[1:pivot+1]之间？
// 对于inorder[pivot+1:], 其值分布在preorder[pivot+1:]
func buildTree(preorder []int, inorder []int) *TreeNode {
	if len(inorder) == 0 {
		return nil
	}
	pivot := findIndex(inorder, preorder[0])

	return &TreeNode{
		preorder[0],
		buildTree(preorder[1:pivot+1], inorder[:pivot]),
		buildTree(preorder[pivot+1:], inorder[pivot+1:]),
	}
}

func findIndex(list []int, target int) int {
	for i, v := range list {
		if v == target {
			return i
		}
	}
	return -1
}

File: 105.从前序和中序遍历序列构造二叉树/buildTree.js
// 老题新做
// 前序遍历 = 中左右 = [根节点, (左侧节点的前序遍历数组), (右侧节点的前序遍历数组)]
// 中序遍历 = 左中右 = [(左侧节点的中序遍历数组), 根节点, (右侧节点的中序遍历数组)]
// 那么容易联想到递归，每次从preorder中去除第一个节点作为根节点，然后在inorder中找到其索引，递归即可
function buildTree(preorder, inorder) {
  if (!preorder.length) {
    return null
  }
  const head = new TreeNode(preorder[0])
  // 找到这个index，相当于说：左侧节点的中序遍历数组有index个
  let index
  for (let i = 0; i < inorder.length; i++) {
    if (inorder[i] === head.val) {
      index = i
    }
  }
  head.left = buildTree(preorder.slice(1, index+1), inorder.slice(0, index))
  head.right = buildTree(preorder.slice(index+1), inorder.slice(index+1))
  return head
}

class TreeNode {
  constructor(val) {
    this.val = val
    this.left = null
    this.right = null
  }
}
